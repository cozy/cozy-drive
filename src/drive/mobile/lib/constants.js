export const SOFTWARE_ID = 'io.cozy.drive.mobile'
export const SOFTWARE_NAME = 'Cozy Drive'
export const APP_STORE_ID = '1224102389'
