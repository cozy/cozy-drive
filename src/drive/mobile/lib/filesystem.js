import { isAndroidApp } from 'cozy-device-helper'
import logger from 'lib/logger'
import { isIOS } from 'cozy-device-helper/dist/platform'

const ERROR_GET_DIRECTORY = 'Error to get directory'
const ERROR_WRITE_FILE = 'Error to write file'
const ERROR_GET_FILE = 'Error to get file'
const COZY_PATH = 'Cozy'
const COZY_FILES_PATH = isIOS() ? 'CozyDrive' : 'Cozy Drive'

export const getRootPath = () =>
  isAndroidApp()
    ? window.cordova.file.externalRootDirectory
    : window.cordova.file.dataDirectory

export const getTemporaryRootPath = () =>
  isAndroidApp()
    ? window.cordova.file.externalCacheDirectory
    : window.cordova.file.cacheDirectory

export const getCozyPath = () => COZY_PATH + '/' + COZY_FILES_PATH + '/'

export const getEntry = path =>
  new Promise((resolve, reject) => {
    window.resolveLocalFileSystemURL(path, resolve, err => {
      logger.error(`${path} could not be resolved: ${err.message}`)
      reject(err)
    })
  })

export const getCozyEntry = () =>
  getEntry(getRootPath() + getCozyPath()).catch(() => createCozyPath())

export const createCozyPath = () =>
  getEntry(getRootPath()).then(entry =>
    getDirectory(entry, COZY_PATH).then(entry =>
      getDirectory(entry, COZY_FILES_PATH)
    )
  )

export const getDirectory = (rootDirEntry, folderName) =>
  new Promise((resolve, reject) => {
    rootDirEntry.getDirectory(folderName, { create: true }, resolve, error => {
      logger.warn(ERROR_GET_DIRECTORY, folderName)
      logger.warn(error)
      reject(ERROR_GET_DIRECTORY)
    })
  })

export const writeFile = (fileEntry, dataObj) =>
  new Promise((resolve, reject) => {
    fileEntry.createWriter(fileWriter => {
      fileWriter.onwriteend = () => {
        resolve(fileEntry)
      }
      fileWriter.onerror = error => {
        logger.warn(ERROR_WRITE_FILE)
        logger.warn(error)
        reject(ERROR_WRITE_FILE)
      }
      fileWriter.write(dataObj)
    })
  })

const saveFile = (dirEntry, fileData, fileName) =>
  new Promise((resolve, reject) => {
    dirEntry.getFile(
      fileName,
      { create: true, exclusive: false },
      fileEntry => {
        writeFile(fileEntry, fileData)
          .then(() => {
            resolve(fileEntry)
          })
          .catch(reject)
      },
      error => {
        logger.warn(ERROR_GET_FILE)
        logger.warn(error)
        reject(ERROR_GET_FILE)
      }
    )
  })

export const openFileWithCordova = (URI, mimetype) =>
  new Promise((resolve, reject) => {
    const callbacks = {
      error: reject,
      success: resolve
    }
    window.cordova.plugins.fileOpener2.open(URI, mimetype, callbacks)
  })

export const deleteOfflineFile = async filename => {
  const entry = await getCozyEntry()
  const fileEntry = await getEntry(`${entry.nativeURL}${filename}`)
  return fileEntry.remove()
}

export const saveFileWithCordova = (fileData, fileName) =>
  getCozyEntry().then(entry => saveFile(entry, fileData, fileName))

export const temporarySave = (file, filename) =>
  getEntry(getTemporaryRootPath()).then(entry =>
    saveFile(entry, file, filename)
  )
/**
 *
 * @param {Blob} blob Binary of the file
 * @param {Object} file io.cozy.files object
 */
export const saveAndOpenWithCordova = (blob, file) => {
  return temporarySave(blob, file.name).then(entry =>
    openFileWithCordova(entry.nativeURL, file.mime)
  )
}
export const getNativeFile = async file => {
  const entry = await getCozyEntry()
  return getEntry(`${entry.nativeURL}${file.id}`)
}

export const openOfflineFile = async file => {
  const fileEntry = await getNativeFile(file)
  return openFileWithCordova(fileEntry.nativeURL, file.mime)
}

export const createTemporaryLocalFile = async (client, file) => {
  const response = await client
    .collection('io.cozy.files')
    .fetchFileContent(file)
  const blob = await response.blob()
  const localFile = await temporarySave(blob, file.name)
  return localFile
}
