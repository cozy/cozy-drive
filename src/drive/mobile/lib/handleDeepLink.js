const PROTOCOL = 'cozydrive://'
const RX = new RegExp('^' + PROTOCOL)
import { setOnboarding } from '../modules/authorization/duck'

export const handleDeeplink = (history, store, url) => {
  const stripped = url.replace(RX, '')

  history.replace('/' + stripped)
  if (stripped.includes('auth?')) {
    const currentLocation = history.getCurrentLocation()
    const { code, state, cozy_url } = currentLocation.query
    store.dispatch(
      setOnboarding({
        code,
        state,
        cozy_url
      })
    )
  }
}
