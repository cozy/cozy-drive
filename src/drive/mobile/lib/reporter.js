/* global __SENTRY_TOKEN__, __DEVELOPMENT__ */
import Raven from 'raven-js'

const PROD_ANALYTICS_URL = `https://${__SENTRY_TOKEN__}@sentry.cozycloud.cc/6`
const DEV_ANALYTICS_URL = `https://${__SENTRY_TOKEN__}@sentry.cozycloud.cc/2`
export const ANALYTICS_URL = __DEVELOPMENT__
  ? DEV_ANALYTICS_URL
  : PROD_ANALYTICS_URL

export const getReporterConfiguration = (isEnabled = false) => ({
  shouldSendCallback: () => isEnabled,
  environment: __DEVELOPMENT__ ? 'development' : 'production'
})

export const configureReporter = enable => {
  Raven.config(ANALYTICS_URL, getReporterConfiguration(enable)).install()
}

export const logException = err => {
  return new Promise(resolve => {
    Raven.captureException(err)
    console.warn('Raven is recording exception')
    console.error(err)
    resolve()
  })
}

const logMessage = (message, serverUrl, level = 'info', force) => {
  return new Promise(resolve => {
    if (force) {
      configureReporter(true)
    }
    Raven.setUserContext = {
      url: serverUrl
    }
    Raven.captureMessage(`[${serverUrl}] ${message}`, {
      level
    })
    if (force) {
      configureReporter(false)
    }
    resolve()
  })
}

export const logInfo = (message, serverUrl, force = false) =>
  logMessage(message, serverUrl, 'info', force)
