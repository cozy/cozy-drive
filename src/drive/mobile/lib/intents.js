import { getEntry } from './filesystem'
import Alerter from 'cozy-ui/react/Alerter'
import { ROOT_DIR_ID } from 'drive/constants/config'
import { uploadFiles } from 'drive/web/modules/navigation/duck'

const getFile = (dirEntry, type = '') =>
  new Promise(resolve => {
    dirEntry.file(file => {
      // window.File is modified by cordova, so we need this trick
      const reader = new FileReader()
      reader.onloadend = function() {
        if (file.type) type = file.type

        const blob = new Blob([new Uint8Array(this.result)], {
          type
        })

        const mimeParts = type.split('/')

        if (mimeParts && mimeParts[1]) {
          const ext = mimeParts[1].replace('jpeg', 'jpg')
          const regex = new RegExp(`.${ext}$`)
          blob.name = regex.test(file.name) ? file.name : `${file.name}.${ext}`
        } else {
          blob.name = file.name
        }

        blob.lastModifiedDate = new Date(file.lastModifiedDate)
        resolve(blob)
      }
      reader.readAsArrayBuffer(file)
    })
  })

const resolveNativePath = path =>
  new Promise((resolve, reject) => {
    window.FilePath.resolveNativePath(path, resolve, err => {
      // eslint-disable-next-line no-console
      console.error(
        `${path} could not be resolved by the plugin: ${err.message}`
      )
      reject(err)
    })
  })

const getFiles = contentFiles =>
  Promise.all(
    contentFiles.map(async contentpath => {
      try {
        const dirEntry = await resolveNativePath(contentpath).then(getEntry)
        const file = await getFile(dirEntry)
        return file
      } catch (err) {
        try {
          // eslint-disable-next-line no-console
          console.warn(
            `Unable to get files with their real filename, let's try another way: ${
              err.message
            }`
          )
          const dirEntry = await getEntry(contentpath)
          const file = await getFile(dirEntry)
          return file
        } catch (err) {
          // eslint-disable-next-line no-console
          console.error(err)
          throw new Error(`Unable to get files: ${err.message}`)
        }
      }
    })
  )

const base64toBlob = (dataURI, type) => {
  const byteString = atob(dataURI)
  const ab = new ArrayBuffer(byteString.length)
  const ia = new Uint8Array(ab)

  for (let i = 0; i < byteString.length; i++) {
    ia[i] = byteString.charCodeAt(i)
  }

  return new Blob([ab], { type })
}

export const intentHandlerAndroid = store => async ({
  extras = 'No extras in intent'
}) => {
  if (extras['android.intent.extra.STREAM']) {
    const contentFiles = Array.isArray(extras['android.intent.extra.STREAM'])
      ? extras['android.intent.extra.STREAM']
      : [extras['android.intent.extra.STREAM']]
    try {
      const files = await getFiles(contentFiles)
      store.dispatch(uploadFiles(files, ROOT_DIR_ID))
    } catch (err) {
      Alerter.info('intents.alert.error')
    }
  }
}

const utiToMime = {
  'com.adobe.pdf': 'application/pdf',
  'com.adobe.postscript': 'application/postscript',
  'com.compuserve.gif': 'image/gif',
  'com.microsoft.bmp': 'image/bmp',
  'com.microsoft.word.doc': 'application/msword',
  'com.microsoft.excel.xls': 'application/vnd.ms-excel',
  'com.microsoft.powerpoint.ppt': 'application/mspowerpoint'
  // see https://developer.apple.com/library/content/documentation/Miscellaneous/Reference/UTIRef/Articles/System-DeclaredUniformTypeIdentifiers.html#//apple_ref/doc/uid/TP40009259-SW1
}

const getMimeTypeFromUTI = utis => {
  const knowUTIs = Object.keys(utiToMime)
  const firstKnownUTI = knowUTIs.find(uti => utis.indexOf(uti) >= 0)

  return firstKnownUTI ? utiToMime[firstKnownUTI] : ''
}

export const intentHandlerIOS = store => async intent => {
  const files = await Promise.all(
    intent.items.map(async item => {
      if (item.uti === 'public.image') {
        const blob = base64toBlob(item.data, item.type)
        blob.name = item.name
        return blob
      } else {
        try {
          const dirEntry = await getEntry(item.data)
          const file = await getFile(dirEntry, getMimeTypeFromUTI(item.utis))

          return file
        } catch (e) {
          return null
        }
      }
    })
  )

  const validFiles = files.filter(file => file)
  if (validFiles.length === 0) {
    Alerter.info('intents.alert.error')
  } else {
    try {
      store.dispatch(uploadFiles(validFiles, ROOT_DIR_ID))
    } catch (err) {
      Alerter.info('intents.alert.error')
    }
  }
}
