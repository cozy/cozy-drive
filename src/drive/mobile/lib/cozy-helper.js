/* global cozy */
import { LocalStorage as Storage } from 'cozy-client-js'
import CozyClient, { StackLink } from 'cozy-client'
import PouchLink from 'cozy-pouch-link'
import { isMobileApp, getDeviceName } from 'cozy-device-helper'
import { SOFTWARE_ID, SOFTWARE_NAME } from './constants'
import { disableBackgroundService } from './background'
import { schema, DOCTYPE_FILES } from 'drive/lib/doctypes'
import appMetadata from 'drive/appMetadata'
import { getRedirectUri } from 'drive/mobile/lib/redirect'
import {
  buildRecentQuery,
  buildDriveQuery,
  buildFolderQuery
} from 'drive/web/modules/queries'

export const getLang = () =>
  navigator && navigator.language ? navigator.language.slice(0, 2) : 'en'

export const getOauthOptions = () => {
  return {
    redirectURI: getRedirectUri(appMetadata.slug),
    softwareID: SOFTWARE_ID,
    clientName: `${SOFTWARE_NAME} (${getDeviceName()})`,
    softwareVersion: appMetadata.version,
    clientKind: 'mobile',
    clientURI: 'https://github.com/cozy/cozy-drive/',
    logoURI:
      'https://github.com/cozy/cozy-drive/raw/master/targets/drive/vendor/assets/oauth-app-icon.png',
    policyURI: 'https://files.cozycloud.cc/cgu.pdf'
  }
}

export const initClient = url => {
  const stackLink = new StackLink()

  const pouchLinkOptions = {
    doctypes: [DOCTYPE_FILES],
    doctypesReplicationOptions: {
      [DOCTYPE_FILES]: {
        strategy: 'fromRemote',
        warmupQueries: [
          buildDriveQuery({
            currentFolderId: 'io.cozy.files.root-dir',
            type: 'directory',
            sortAttribute: 'name',
            sortOrder: 'asc'
          }),
          buildRecentQuery(),
          buildFolderQuery('io.cozy.files.root-dir')
        ]
      }
    },
    initialSync: true
  }

  if (isMobileApp()) {
    pouchLinkOptions.pouch = {
      plugins: [require('pouchdb-adapter-indexeddb').default],
      options: {
        adapter: 'indexeddb',
        location: 'default'
      }
    }
  }

  const pouchLink = new PouchLink(pouchLinkOptions)

  return new CozyClient({
    uri: url,
    oauth: getOauthOptions(),
    appMetadata,
    schema,
    links: [pouchLink, stackLink],
    store: false
  })
}

export const initBar = async client => {
  // Prevents the bar to be initialized 2 times in a row after the onboarding
  if (document.getElementById('coz-bar')) {
    return
  }
  await cozy.bar.init({
    appName: 'Drive',
    appNamePrefix: 'Cozy',
    appSlug: 'drive',
    cozyClient: client,
    iconPath: require('../../targets/vendor/assets/app-icon.svg'),
    lang: getLang(),
    cozyURL: client.uri,
    replaceTitleOnMobile: false,
    displayOnMobile: true
  })
}

export const restoreCozyClientJs = (uri, clientInfos, token) => {
  const offline = { doctypes: ['io.cozy.files'] }
  cozy.client.init({
    cozyURL: uri,
    oauth: {
      storage: new Storage(),
      clientParams: {
        ...clientInfos,
        scopes: token.scope
      }
    },
    offline
  })

  cozy.client.saveCredentials(clientInfos, token)
}

export function resetClient(client, clientInfo = null) {
  if (clientInfo && cozy.client.auth.unregisterClient) {
    cozy.client.auth.unregisterClient(clientInfo)
  }
  // reset cozy-bar
  if (document.getElementById('coz-bar')) {
    document.getElementById('coz-bar').remove()
  }
  // reset pouchDB
  if (cozy.client.offline.destroyAllDatabase) {
    cozy.client.offline.destroyAllDatabase()
  }
  // reset cozy-client
  client.getStackClient().resetClient()
  // reset cozy-client-js
  if (cozy.client._storage) {
    cozy.client._storage.clear()
  }

  disableBackgroundService()
}

export const getToken = async () => {
  const credentials = await cozy.client.authorize()
  return credentials.token.accessToken
}

export const getClientUrl = () => cozy.client._url
