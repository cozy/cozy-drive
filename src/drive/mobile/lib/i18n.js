import { getLang } from './cozy-helper'
import {
  _polyglot,
  initTranslation
} from 'cozy-ui/transpiled/react/I18n/translation'

// used when translating things outside of components, eg. native UIs
export const getTranslateFunction = () => {
  if (_polyglot === undefined) {
    const lang = getLang()
    const dictRequire = lang => require(`../../locales/${lang}`)
    initTranslation(lang, dictRequire)
  }
  return _polyglot.t.bind(_polyglot)
}
