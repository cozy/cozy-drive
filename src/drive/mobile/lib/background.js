/* eslint no-console: 0 */
import configureStore from 'drive/store/configureStore'
import { loadState } from 'drive/store/persistedState'
import { startMediaBackup } from 'drive/mobile/modules/mediaBackup/duck'
import { initClient } from './cozy-helper'
import { logException, configureReporter } from 'drive/lib/reporter'
import { getTranslateFunction } from './i18n'

import {
  getPlatform,
  isMobileApp,
  isIOSApp,
  isAndroidApp
} from 'cozy-device-helper'

/*
  This files is split on 4 parts:
    - GENERIC: Platform agnostic
    - SERVICE: The service background
    - ANDROID: The Android-specific part
    - IOS: The iOS-specific part

  Only 3 functions are export:
    - updateStatusBackgroundService: to enbale/disable background service
    - startBackgroundService: to start background service
    - disableBackgroundService: to disable background service

  For each platform (Android, iOS) we have:
    - has__platform__CordovaPlugin: Checks if the plugin is installed
    - enable__platform__BackgroundService and disable__platform__BackgroundService
    - start__platform__BackgroundService: Allowing to add a specific callback at the end of the service
*/

// GENERIC

export const updateStatusBackgroundService = backupImages => {
  if (backupImages) {
    enableBackgroundService()
  } else {
    disableBackgroundService()
  }
}

export const startBackgroundService = () => {
  if (hasIosCordovaPlugin()) {
    startIosBackgroundService()
  } else if (hasAndroidCordovaPlugin()) {
    startAndroidBackgroundService()
  } else {
    notCompatibleError()
  }
}

const enableBackgroundService = () => {
  if (hasIosCordovaPlugin()) {
    enableIosBackgroundService()
  } else if (hasAndroidCordovaPlugin()) {
    enableAndroidBackgroundService()
  } else {
    notCompatibleError()
  }
}

export const disableBackgroundService = async () => {
  if (hasIosCordovaPlugin()) {
    disableIosBackgroundService()
  } else if (hasAndroidCordovaPlugin()) {
    await disableAndroidBackgroundService()
  } else {
    notCompatibleError()
  }
}

const notCompatibleError = () => {
  const msg = 'Background Service is not compatible with your platform.'
  console.warn(msg)
  if (isMobileApp()) {
    console.log(getPlatform())
    if (isIOSApp()) {
      console.log(window.BackgroundFetch)
    } else if (isAndroidApp()) {
      console.log(window.JSBackgroundService)
    }
  }
  logException(msg)
}

// SERVICE

const backgroundService = () =>
  new Promise(resolve => {
    console.log('BackgroundFetch initiated')

    loadState()
      .then(persistedState => {
        const cozyURL = persistedState.mobile.settings.serverUrl
        configureReporter()
        const client = initClient(cozyURL)
        const store = configureStore(
          client,
          getTranslateFunction(),
          persistedState
        )
        return store.dispatch(startMediaBackup())
      })
      .then(resolve)
      .catch(resolve)
  })

// ANDROID

const hasAndroidCordovaPlugin = () =>
  isAndroidApp() && window.JSBackgroundService !== undefined

const enableAndroidBackgroundService = async () => {
  const isEnable = await isEnableAndroidBackgroundService()
  if (!isEnable) {
    const repeatingPeriod = 60 * 60 * 1000
    window.JSBackgroundService.setRepeating(repeatingPeriod, err => {
      if (err) {
        console.warn(err)
        logException('enableAndroidBackgroundService error')
      }
    })
  }
}

const disableAndroidBackgroundService = async () => {
  const isEnable = await isEnableAndroidBackgroundService()
  if (isEnable) {
    window.JSBackgroundService.cancelRepeating(err => {
      if (err) {
        console.warn(err)
        logException('disableAndroidBackgroundService error')
      }
    })
  }
}

const startAndroidBackgroundService = () => {
  backgroundService().then(() => {
    stopAndroidBackgroundService()
  })
}

const stopAndroidBackgroundService = () => {
  window.service.workDone()
}

const isEnableAndroidBackgroundService = () =>
  new Promise(resolve => {
    window.JSBackgroundService.isRepeating((err, isSet) => {
      if (err) {
        console.warn(err)
        logException('isEnableAndroidBackgroundService error')
        resolve(false)
      }
      resolve(isSet)
    })
  })

// IOS

const hasIosCordovaPlugin = () =>
  isIOSApp() && window.BackgroundFetch !== undefined

const enableIosBackgroundService = () => {
  // documentation: https://github.com/transistorsoft/cordova-plugin-background-fetch
  const fetcher = window.BackgroundFetch

  const failureCallback = error => {
    console.log('BackgroundFetch failed', error)
  }

  const options = {
    stopOnTerminate: false
  }
  fetcher.configure(startIosBackgroundService, failureCallback, options)
}

const disableIosBackgroundService = () => {
  window.BackgroundFetch.stop()
}

const startIosBackgroundService = () => {
  backgroundService().then(() => {
    stopIosBackgroundService()
  })
}

const stopIosBackgroundService = () => {
  window.BackgroundFetch.finish()
}
