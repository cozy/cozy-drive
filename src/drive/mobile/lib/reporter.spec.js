import { normalizeData } from 'drive/lib/reporter'

describe('Normalization function', () => {
  it('should normalize the culprit', () => {
    const data = { culprit: 'file://app.js' }
    expect(normalizeData(data)).toMatchSnapshot()
  })

  it('should normalize stacktraces', () => {
    const stacktrace = {
      frames: [
        { filename: 'file://app.js', random: 5 },
        { filename: 'file:///android_asset/www/app.js' },
        {
          filename:
            'file:///var/containers/Bundle/Application/1234-5678-ABCD/Cozy%20Drive.app/www/app.js'
        },
        { filename: '[native code]' }
      ]
    }
    const data = {
      exception: {
        values: [{ stacktrace }]
      }
    }

    expect(normalizeData(data)).toMatchSnapshot()
  })
})
