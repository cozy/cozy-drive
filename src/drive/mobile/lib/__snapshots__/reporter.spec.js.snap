// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normalization function should normalize stacktraces 1`] = `
Object {
  "exception": Object {
    "values": Array [
      Object {
        "stacktrace": Object {
          "frames": Array [
            Object {
              "filename": "app://app.js",
              "random": 5,
            },
            Object {
              "filename": "/app.js",
            },
            Object {
              "filename": "/app.js",
            },
            Object {
              "filename": "[native code]",
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`Normalization function should normalize the culprit 1`] = `
Object {
  "culprit": "app://app.js",
}
`;
